pallets/did/src/lib.rs

```diff
@@ -28,93 +28,364 @@ mod tests;
+	// 		DIDEncryptionKeyType::KeyAgreement => Option::from(&self.key_agreement_key),
+	// 	}
+	// }
+}
+
+/// A trait describing an operation that requires DID authentication.
+pub trait DIDOperation: Encode {
+	/// Returns the type of the verification key to be used to validate the operation.
+	fn get_verification_key_type(&self) -> DIDVerificationKeyType;
+	/// Returns the DID identifier of the subject.
+	fn get_did(&self) -> &DIDIdentifier;
+}
+
+/// An enum describing the different verification methods a verification key can fulfil, according to the [DID specification](https://w3c.github.io/did-spec-registries/#verification-relationships).
+#[derive(Clone, Debug, Decode, Encode, PartialEq)]
+pub enum DIDVerificationKeyType {
+	Authentication,
+	CapabilityDelegation,
+	CapabilityInvocation,           // For future use
+	AssertionMethod,
+}
+
+/// An enum describing the different verification methods an encryption key can fulfil, according to the [DID specification](https://w3c.github.io/did-spec-registries/#verification-relationships).
+#[derive(Clone, Debug, Decode, Encode, PartialEq)]
+pub enum DIDEncryptionKeyType {
+	KeyAgreement,
+}
 
-	/// Public boxing key type for DIDs
-	type PublicBoxKey: Parameter + Member + Codec + Default;
+/**
+A DID creation request. It contains the following values:
+- the DID identifier being created;
+- the new authentication key to use;
+- the new encryption key to use;
+- the optional attestation key to use;
+- the optional delegation key to use;
+- the optional endpoint URL pointing to the DID service endpoints.
+ */
+#[derive(Clone, Decode, Debug, Encode, PartialEq)]
+pub struct DIDCreationOperation {
+	did: DIDIdentifier,
+	new_auth_key: PublicVerificationKey,
+	new_key_agreement_key: PublicEncryptionKey,
+	new_attestation_key: Option<PublicVerificationKey>,
+	new_delegation_key: Option<PublicVerificationKey>,
+	new_endpoint_url: Option<URL>,
+}
+
+impl DIDOperation for DIDCreationOperation {
+    fn get_verification_key_type(&self) -> DIDVerificationKeyType {
+        DIDVerificationKeyType::Authentication
+    }
+
+    fn get_did(&self) -> &DIDIdentifier {
+        self.did.as_ref()
+    }
+}
+
+/// All the errors that can be generated when evaluating a DID operation.
+pub enum DIDError {
+	/// The signature is not in the expected format the verification key expects.
+	InvalidSignatureFormat,
+	/// The signature is invalid for the payload and the verification key provided.
+	InvalidSignature,
+	/// The DID being created is already present on chain.
+	DIDAlreadyPresent,
+	/// The expected DID cannot be found on chain.
+	DIDNotPresent,
+	/// The given DID does not contain the right key to verify the signature of a DID operation.
+	VerificationkeyNotPresent(DIDVerificationKeyType),
+}
+
+// Used internally to handle signature errors.
+pub enum SignatureError {
+	InvalidSignatureFormat,
+	InvalidSignature,	
+}
+
+decl_error! {
+	// TODO: How to add the DIDError enum to here and use within the extrinsics?
```

Comment:
How can we have nested errors exposed to the runtime? I wanted to split errors in signature errors, DID errors, and storage errors, but then did not manage to build compiler-compliant error objects from within the extrinsic...